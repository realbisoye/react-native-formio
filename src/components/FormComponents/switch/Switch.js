import React from 'react';
import {View, Switch, Text, StyleSheet} from 'react-native';
import ValueComponent from '../sharedComponents/Value';

export default class SwitchComponent extends ValueComponent {

  constructor(props) {
    super(props);

    this.onChange = this.onChange.bind(this);
    this.getElements = this.getElements.bind(this);
    this.getValueDisplay = this.getValueDisplay.bind(this);
  }

  onChange(checked) {
    this.setValue(checked);
  }

  getElements() {
    const switchStyle = StyleSheet.create({
      wrapper: {
        flex: 1,
        marginTop: 10,
        marginLeft: 20,
        flexDirection: 'row'
      },
      label: {
        fontSize: 15,
        marginTop: 8,
        marginLeft: 10,
        color: this.props.colors.textColor,
      },
    });

    const {component} = this.props;
    const {value} = this.state;

    return (
      <View style={switchStyle.wrapper}>
        <Switch
          tintColor={this.props.colors.primary1Color}
          value={value.item}
          disabled={this.props.readOnly}
          onValueChange={this.onChange}
        />
        <Text style={switchStyle.label}>
          {!(component.hideLabel && component.datagridLabel === false) ? component.label : ''}
        </Text>
      </View>
    );
  }

  getValueDisplay(component, data) {
    return data ? 'Yes' : 'No';
  }
}
