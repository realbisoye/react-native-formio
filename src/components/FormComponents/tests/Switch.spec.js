import React from 'react';
import renderer from 'react-test-renderer';
import colors from '../../../defaultTheme/colors';
import Switch from '../switch/Switch';

describe('Switch', () => {
  describe('Switch field', () => {
    const component= {
      'conditional': {
        'eq': '',
        'when': null,
        'show': ''
      },
      'type': 'checkbox',
      'validate': {
        'required': false
      },
      'persistent': true,
      'protected': false,
      'defaultValue': false,
      'key': 'checkbox',
      'label': 'Checkbox',
      'hideLabel': true,
      'tableView': true,
      'inputType': 'checkbox',
      'input': true
    };
    var attachToForm = jest.fn();
    it.only('Renders a checkbox field', () => {
      const element = renderer.create(
        <Switch
          colors={colors}
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element).toMatchSnapshot();
    });

    it('Renders a checkbox field without a label', () => {
      component.hideLabel = true;
      component.datagridLabel = false;
      const element = renderer.create(
        <Switch
          colors={colors}
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element.find('.checkbox label').html()).to.equal('<input type="checkbox" id="checkbox">');
      delete component.datagridLabel;
    });

    it('Renders a checkbox field with a label when variables set', () => {
      component.hideLabel = false;
      component.datagridLabel = false;
      const element = renderer.create(
        <Switch
          colors={colors}
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element.find('.checkbox label').html()).to.equal('<input type="checkbox" id="checkbox">Checkbox');
      delete component.datagridLabel;
      component.hideLabel = true;
    });

    it('Renders a checkbox field with a label when variables set', () => {
      component.hideLabel = true;
      component.datagridLabel = true;
      const element = renderer.create(
        <Switch
          colors={colors}
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element.find('.checkbox label').html()).to.equal('<input type="checkbox" id="checkbox">Checkbox');
      delete component.datagridLabel;
      component.hideLabel = true;
    });

    it('Renders a checkbox field with a label when variables set', () => {
      component.hideLabel = false;
      component.datagridLabel = true;
      const element = renderer.create(
        <Switch
          colors={colors}
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element.find('.checkbox label').html()).to.equal('<input type="checkbox" id="checkbox">Checkbox');
      delete component.datagridLabel;
      component.hideLabel = true;
    });

    it('Sets a default value as true', () => {
      component.defaultValue = true;
      const element = renderer.create(
        <Switch
          colors={colors}
          component={component}
          value={component.defaultValue}
          attachToForm={attachToForm}
        />
      );
      expect(element.attr('checked')).to.equal('checked');
      component.defaultValue = false;
    });

    it('Check the validations', () => {
      component.validate.required = true;
      const element = renderer.create(
        <Switch
          colors={colors}
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element.find('.checkbox label').attr('class')).to.equal('control-label field-required not-checked');
    });

    it('Sets the checked class when selected', () => {
      const element = renderer.create(
        <Switch
          colors={colors}
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element.find('label').hasClass('not-checked')).to.equal(true);
      element.find('input').simulate('change', {'target': {'checked': true}});
      expect(element.find('label').hasClass('checked')).to.equal(true);
    });

    it('sets a custom class', () => {
      component.customClass = 'my-custom-class';
      const element = renderer.create(
        <Switch
          colors={colors}
          component={component}
          attachToForm={attachToForm}
        />
      );
      expect(element.attr('class').split(' ')).to.contain('my-custom-class');
    });
  });
});
