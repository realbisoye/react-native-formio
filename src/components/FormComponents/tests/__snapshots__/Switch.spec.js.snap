// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Switch field Renders a checkbox field 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
      "marginLeft": 20,
      "marginTop": 10,
    }
  }
>
  <RCTSwitch
    disabled={false}
    onChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onValueChange={[Function]}
    style={
      Object {
        "height": 31,
        "width": 51,
      }
    }
    tintColor="#03a9f4"
    value={false}
  />
  <Text
    style={
      Object {
        "color": "rgba(0, 0, 0, 0.87)",
        "fontSize": 15,
        "marginLeft": 10,
        "marginTop": 8,
      }
    }
  >
    Checkbox
  </Text>
</View>
`;
